core/metrics.go

```diff
@@ -15,199 +15,147 @@
-}
+	MetricsRegistry = metrics.NewRegistry()
+	metricsStreamRegistry = make(map[MessageStreamID]*StreamMetric)
+
+	metricVersion = metrics.NewRegisteredGauge("version", MetricsRegistry)
+	metricBuild = metrics.NewRegisteredGauge("build", MetricsRegistry)
+	metricGoVersion = metrics.NewRegisteredGauge("go_version", MetricsRegistry)
+	MetricMessagesRouted = metrics.NewRegisteredCounter("routed", MetricsRegistry)
+	MetricMessagesEnqued = metrics.NewRegisteredCounter("enqueued", MetricsRegistry)
+	MetricMessagesDiscarded = metrics.NewRegisteredCounter("discarded", MetricsRegistry)
+	MetricActiveWorkers = metrics.NewRegisteredCounter("workers", MetricsRegistry)
+
+	pluginMetricsRegistry = NewSubRegistry("plugins")
+	MetricRouters = metrics.NewRegisteredCounter("routers", pluginMetricsRegistry)
+	MetricFallbackRouters = metrics.NewRegisteredCounter("routers_default", pluginMetricsRegistry)
+	MetricConsumers = metrics.NewRegisteredCounter("consumers", pluginMetricsRegistry)
+	MetricProducers = metrics.NewRegisteredCounter("producers", pluginMetricsRegistry)
+	MetricPluginsInit = metrics.NewRegisteredCounter("init", pluginMetricsRegistry)
+	MetricPluginsWaiting = metrics.NewRegisteredCounter("waiting", pluginMetricsRegistry)
+	MetricPluginsActive = metrics.NewRegisteredCounter("active", pluginMetricsRegistry)
+	MetricPluginsPrepareStop = metrics.NewRegisteredCounter("preparestop", pluginMetricsRegistry)
+	MetricPluginsStopping = metrics.NewRegisteredCounter("stopping", pluginMetricsRegistry)
+	MetricPluginsDead = metrics.NewRegisteredCounter("dead", pluginMetricsRegistry)
+
+	stateToMetric[PluginStateInitializing] = MetricPluginsInit
+	stateToMetric[PluginStateWaiting] = MetricPluginsWaiting
+	stateToMetric[PluginStateActive] = MetricPluginsActive
+	stateToMetric[PluginStatePrepareStop] = MetricPluginsPrepareStop
+	stateToMetric[PluginStateStopping] = MetricPluginsStopping
+	stateToMetric[PluginStateDead] = MetricPluginsDead
+
+	metrics.RegisterRuntimeMemStats(MetricsRegistry)
+
+	// Populate constant values
+
+	version := runtime.Version()
+	if version[0] == 'g' && version[1] == 'o' {
+		parts := strings.Split(version[2:], ".")
+		numericVersion := make([]uint64, tmath.MaxI(3, len(parts)))
+		for i, p := range parts {
+			numericVersion[i], _ = strconv.ParseUint(p, 10, 64)
+		}
+		metricGoVersion.Update(int64(numericVersion[0]*10000 + numericVersion[1]*100 + numericVersion[2]))
+	}
 
-// CountConsumers increases the consumer counter by 1
-func CountConsumers() {
-	tgo.Metric.Inc(metricCons)
+	ver, build := GetVersionNumber()
+	metricVersion.Update(ver)
+	metricBuild.Update(int64(build))
+	go metrics.CaptureRuntimeMemStats(MetricsRegistry, time.Second)
 }
 
-// CountRouters increases the stream counter by 1
-func CountRouters() {
-	tgo.Metric.Inc(metricRouters)
+// NewSubRegistry creates a new, prefixed metrics registry that can be used
+// to register custom plugin metrics.
+func NewSubRegistry(prefix string) metrics.Registry {
```

Comment:
Could NewSubRegistry() and NewPluginRegistry() be renamed to something like NewMetricsSubRegistry() and NewMetricsPluginRegistry()? Currently when you encounter references to these elsewhere in the code, nothing tells you that they're metrics-specific functions (other than the return value's type, which is invisible) - they look like central parts of Gollum's core instead.

Reply:
good point

Reply:
adjusted. It's now `NewMetricsRegistry` and `NewMetricsRegistryForPlugin`.